src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -939,7 +944,43 @@ private static string EncryptTokenPrivate(string innerJwt, EncryptingCredentials
             }
         }
 
-        internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken, TokenValidationParameters validationParameters)
+        private static SecurityKey ResolveTokenDecryptionKeyFromConfig(JsonWebToken jwtToken, BaseConfiguration configuration)
+        {
+            if (jwtToken == null)
+                throw LogHelper.LogArgumentNullException(nameof(jwtToken));
+
+            if (!string.IsNullOrEmpty(jwtToken.Kid))
+            {
+                if (configuration.TokenDecryptionKeys != null)
+                {
+                    foreach (var key in configuration.TokenDecryptionKeys)
+                    {
+                        if (key != null && string.Equals(key.KeyId, jwtToken.Kid, GetStringComparisonRuleIf509OrECDsa(key)))
+                            return key;
+                    }
+                }
+            }
+
+            if (!string.IsNullOrEmpty(jwtToken.X5t))
+            {
+                if (configuration.TokenDecryptionKeys != null)
+                {
+                    foreach (var key in configuration.TokenDecryptionKeys)
+                    {
+                        if (key != null && string.Equals(key.KeyId, jwtToken.X5t, GetStringComparisonRuleIf509(key)))
+                            return key;
+
+                        var x509Key = key as X509SecurityKey;
+                        if (x509Key != null && string.Equals(x509Key.X5t, jwtToken.X5t, StringComparison.OrdinalIgnoreCase))
+                            return key;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
         {
             IEnumerable<SecurityKey> keys = null;
 
```

Comment:
Added below comments to describe the logic and log messages to know where the key was found. 

Since we are already finding the key that matches kid, that is our key wherever it is found. so IMO we need not continue to check others(eg: config).

// First we check to see if the caller has set a custom decryption resolver on TVP for the call, if so any keys set on TVP and keys in Configuration are ignored.
 // If no custom decryption resolver set, we'll check to see if they've set some static decryption keys on TVP. If a key found, we ignore configuration.
 // If no key found in TVP, we'll check the configuration.